File: 0017-letter-combinations-of-a-phone-number/solution.cpp
class Solution {
    string buttons[8]={"abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
    vector<string>v;
public:
    void gen(string digits, int idx, string pre=""){
        if(idx>=digits.size())v.push_back(pre);
        else{
            int button=digits[idx]-'0'-2;
            for(int i=0;i<buttons[button].size();i++){
                gen(digits,idx+1,pre+buttons[button][i]);
            }
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.size())gen(digits,0);
        return v;
    }
};

File: 0017-letter-combinations-of-a-phone-number/solution.java
class Solution {
    char[][] mp = new char[][]{
        {},
        {},
        {'a','b','c'},
        {'d','e','f'},
        {'g','h','i'},
        {'j','k','l'},
        {'m','n','o'},
        {'p','q','r','s'},
        {'t','u','v'},
        {'w','x','y','z'},
    };
    LinkedList<String> res = new LinkedList<>();
    char[] digits;
    char[] curr;

    void goBabe(int idx){
        if(idx==digits.length){
            res.add(new String(curr));
        }else{
            int digit = digits[idx] - '0';
            for(char c : mp[digit]){
                curr[idx] = c;
                goBabe(idx+1);
            }
        }
    }

    public List<String> letterCombinations(String digits) {
        if((digits==null)||(digits.length()==0)){
            return res;
        }
        curr = new char[digits.length()];
        this.digits = digits.toCharArray();
        goBabe(0);
        return res;
    }
}
